api/metric/api.go

```diff
@@ -21,62 +23,231 @@ import (
 	"go.opentelemetry.io/api/unit"
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid     Kind = iota
+	CounterKind      // Supports Add()
+	GaugeKind        // Supports Set()
+	MeasureKind      // Supports Record()
+	ObserverKind
 )
 
-type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Handle interface {
+	// Record allows the SDK to observe a single metric event
+	RecordOne(ctx context.Context, value MeasurementValue)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+// LabelSet represents a []core.KeyValue for use as pre-defined labels
+// in the metrics API.
+//
+// TODO this belongs outside the metrics API, in some sense, but that
+// might create a dependency.  Putting this here means we can't re-use
+// a LabelSet between metrics and tracing, even when they are the same
+// SDK.
+type LabelSet interface {
+	Meter() Meter
 }
 
-type Handle struct {
-	Name        string
-	Description string
-	Unit        unit.Unit
+// ObserverCallback defines a type of the callback SDK will call for
+// the registered observers.
+type ObserverCallback func(Meter, Observer) (LabelSet, MeasurementValue)
+
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
+type Meter interface {
+	// DefineLabels returns a reference to a set of labels that
+	// cannot be read by the application.
+	DefineLabels(context.Context, ...core.KeyValue) LabelSet
+
+	NewHandle(context.Context, Descriptor, LabelSet) Handle
```

Comment:
Let's remove the `context.Context` argument here. It could cause confusion -> the Handle doesn't take any tags from the context. I prefer to use Context everywhere, as a sort of future-proofing, but in this case I think it will lead to confusion.

Reply:
Will do.